### PR TITLE
"Source" tag for metrics

### PR DESCRIPTION
**Pull request by s0nicboOm**
1. Add source tags for the metrics
2. Fix some logs

### CODE REVIEW

File: numalogic/monitoring/metrics.py [hunk 1]
```diff
@@ -92,3 +92,28 @@ def add_observation(self, *label_values, value: float) -> None:
         if len(label_values) != len(self.label_keys):
             raise ValueError(f"Labels mismatch with the definition: {self.label_keys}")
         self.summary.labels(*label_values).observe(amount=value)
+
+
+class PromGaugeMetric(_BaseMetric):
+    """Class is used to create an info object and increment it."""
+
+    __slots__ = "info"
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:100]:
```suggestion
    __slots__ = ("info",)
```

File: numalogic/monitoring/metrics.py [hunk 2]
```diff
@@ -92,3 +92,28 @@ def add_observation(self, *label_values, value: float) -> None:
...
+
+    def __init__(self, name: str, description: str, label_keys: Optional[list[str]]) -> None:
+        super().__init__(name, description, label_keys)
+        self.info = Gauge(name, description, label_keys)
+
+    def set_gauge(
+        self,
+        *label_values,
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:108]:
```suggestion
        *label_values: str,
```

File: numalogic/monitoring/metrics.py [hunk 3]
```diff
@@ -92,3 +92,28 @@ def add_observation(self, *label_values, value: float) -> None:
...
+        data: float,
+    ) -> None:
+        """
+        Utility function is used to increment the info.
+        Args:
+            *label_values: List of labels
+            data: float data.
+        """
+        if len(label_values) != len(self.label_keys):
+            raise ValueError(f"Labels mismatch with the definition: {self.label_keys}")
```
Comment (ab93) [inline numalogic/monitoring/metrics.py:118]:
```suggestion
            raise ValueError(f"Labels length mismatch with the definition: {self.label_keys}")
```

File: numalogic/udfs/tools.py [hunk 4]
```diff
@@ -57,6 +60,24 @@ def get_df(
...
+        metric_name: metric name in the payload
+        labels: labels.
+    """
+    for _data, _metric_name in zip(data.T, metric_name):
+        _set_gauge(
+            gauge=RECORDED_DATA_GAUGE,
+            labels=(*labels, _metric_name),
+            data=np.mean(_data).squeeze(),
+        )
+
```
Comment (ab93) [inline numalogic/udfs/tools.py:79]: Can you explain what are we trying to do here?

File: numalogic/udfs/preprocess.py [hunk 5]
```diff
@@ -99,9 +100,15 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
         _stream_conf = self.get_stream_conf(data_payload["config_id"])
         _conf = _stream_conf.ml_pipelines[data_payload["pipeline_id"]]
         raw_df, timestamps = get_df(data_payload=data_payload, stream_conf=_stream_conf)
-
-        # TODO: Add pipeline id to the ml metrics
+        try:
+            source = data_payload["metadata"]["numalogic_opex_tags"]["source"]
+        except Exception:
```
Comment (ab93) [inline numalogic/udfs/preprocess.py:105]: Wont just catching KeyError work?

File: numalogic/udfs/tools.py (hunk 4 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:79]: This basically takes the mean of the data metric wise.

File: numalogic/udfs/preprocess.py (hunk 5 above)
Comment (s0nicboOm) [reply numalogic/udfs/preprocess.py:105]: We will otherwise have to check existence of keys: meatadata, numalogic_opex_tags and source.

Comment (s0nicboOm) [reply numalogic/udfs/preprocess.py:105]: What do you think?

File: numalogic/monitoring/metrics.py (hunk 1 above)
Comment (s0nicboOm) [reply numalogic/monitoring/metrics.py:100]: I think ruff removes the bracket saying it is redundant

File: numalogic/udfs/tools.py [hunk 6]
```diff
@@ -57,6 +60,25 @@ def get_df(
...
+    data: np.ndarray, metric_names: Sequence[str], labels: Sequence[str]
+) -> None:
+    """
+    Utility function is used to update the gauge metric.
+    Args:
+        data: data
+        metric_names: metric name in the payload
+        labels: labels.
+    """
+    metric_mean = np.mean(data.reshape(-1, 1), axis=0)
```
Comment (ab93) [inline numalogic/udfs/tools.py:73]: Why reshape here?

Comment (ab93) [inline numalogic/udfs/tools.py:73]: Throw an error if len(metric_names) is not equal to num columns in data

File: numalogic/udfs/tools.py [hunk 7]
```diff
@@ -88,7 +110,7 @@ def make_stream_payload(
 # TODO: move to base NumalogicUDF class and look into payload mutation
 def _get_artifact_stats(artifact_data):
     return {
-        "source": None or artifact_data.extras.get("source"),
+        "artifact_source": None or artifact_data.extras.get("source"),
```
Comment (ab93) [inline numalogic/udfs/tools.py:113]: Why is `None` checked before the real data?

File: numalogic/udfs/tools.py [hunk 8]
```diff
@@ -270,6 +304,7 @@ def ack_read(
         Acknowledge the read message. Return True when the msg has to be trained.
         Args:
             key: key
+            pipeline_id: key
```
Comment (ab93) [inline numalogic/udfs/tools.py:307]: Here also can we do like we discussed?

File: numalogic/udfs/tools.py (hunk 6 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:73]: Yes

File: numalogic/udfs/tools.py (hunk 8 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:307]: Let me remove it from the docstring. I might have missed to remove it.

File: numalogic/udfs/tools.py (hunk 6 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:73]: postprocess sends the value after taking a mean and we get it like [a, b] for 2 let's say [falied, degraded]

File: numalogic/udfs/tools.py (hunk 7 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:113]: fixed!

**Review by ab93 — APPROVED** [review]

**Pull request merged**